content/post/2020-01-15-examples/index.Rmd

```diff
@@ -0,0 +1,148 @@
+
+## How to safeguard examples?
+
+Examples are supposed to be _executable_ i.e. to run without error. As writing in ["Writing R Extensions"](https://cran.r-project.org/doc/manuals/R-exts.html#index-_005cdontrun), _"By default, text inside `\examples{}` will be displayed in the output of the help page and run by `example()` and by `R CMD check`."_. `R CMD check`'s running example could be described as a [_smoke test_](https://en.wikipedia.org/wiki/Smoke_testing_(software)).
+
+Now comes one of the trickiest aspects of creating examples for the manual pages: how to execute them only when they work? We're not encouraging cheating but
+
+* Examples can show an error, to exemplify the behaviour of the function in such cases.
+* Examples can need user setup such as storing an API key in `.Renviron`.
+* Examples can be too long for `R CMD check`.
+
+Thankfully there are few toggles that you can use for your examples. From "Writing R extensions", _"You can use `\dontrun{}` for text that should only be shown, but not run, and `\dontshow{}` for extra commands for testing that should not be shown to users"_ and _"Finally, there is `\donttest`, used (at the beginning of a separate line) to mark code that should be run by example() but not by R CMD check (by default: the option `--run-donttest` can be used)."_. In the help files the user reads, `\dontshow{}` lines are well, not shown; `dontrun{}` blocks come between `## Not run:` and `## End(Not run)` lines; `\donttest{}` lines look like all other lines. (See ` lookup::lookup(tools::Rd2HTML)`) That's a lot of information to process, but `roxygen2` docs feature [a handy reference table](https://roxygen2.r-lib.org/articles/rd.html#functions).
+
+To complicate things a bit more, `example()` and `R CMD check` have toggles too! You could run `example(<topic>, run.dontrun = TRUE)` and `R CMD check` with the `--run-donttest` options. :dizzy_face:
+
+Here's a summary
+
+* Example that works out of the box but long or interactive :arrow_right: donttest (used [2,019 times in CRAN packages at the time of writing](https://github.com/search?utf8=%E2%9C%93&q=%22%5Cdontrun%7B%22+user%3Acran+extension%3ARd&type=Code&ref=advsearch&l=&l=))
+
+* Example that doesn't work out of the box or doesn't work at all :arrow_right: dontrun (used [19,065 times in CRAN packages at the time of writing](https://github.com/search?utf8=%E2%9C%93&q=%22%5Cdontrun%7B%22+user%3Acran+extension%3ARd&type=Code&ref=advsearch&l=&l=))
+
+> For a CRAN submission you might be told to use `donttest` for examples using a token.
+
+* Example where you want to [sneakily a subset of the example data](https://kbroman.org/pkg_primer/pages/cran.html) (to make it run faster) :arrow_right: dontshow (used [1,209 times in CRAN packages at the time of writing](https://github.com/search?q=%22%5Cdontshow%7B%22+user%3Acran+extension%3ARd&type=Code))
```

Comment:
> sneakily a subset of the example data

sneakily subset the example data

But I am not totally sure if this is a good idea TBH, because you get different results with `example()` and simple copy-and-paste. I don't think many people are using `example()` but we do in this post....

Maybe just mention this caveat?